crates/matrix-sdk-crypto/src/store/memorystore.rs

```diff
@@ -272,47 +340,90 @@ impl CryptoStore for MemoryStore {
+            if let Some(backup_order) = backup_version.and_then(|v| self.get_backup_order(v)) {
+                self.get_inbound_group_sessions_and_backup_order()
+                    .await?
+                    .into_iter()
+                    .filter(|(_, o)| *o == backup_order)
+                    .count()
+            } else {
+                0
+            };
 
         Ok(RoomKeyCounts { total: self.inbound_group_sessions.count(), backed_up })
     }
 
     async fn inbound_group_sessions_for_backup(
         &self,
-        _backup_version: &str,
+        backup_version: &str,
         limit: usize,
     ) -> Result<Vec<InboundGroupSession>> {
-        Ok(self
-            .get_inbound_group_sessions()
-            .await?
-            .into_iter()
-            .filter(|s| !s.backed_up())
-            .take(limit)
-            .collect())
+        // Given an iterator, limit it, collect it into a `Vec` and wrap it in `Ok`.
+        // The two branches of the `if` below both result in iterators, but because they
+        // are different types it is difficult to collect them into an
+        // intermediate variable, so we share code using this function instead.
+        fn take_and_collect<T: Iterator<Item = InboundGroupSession>>(
+            limit: usize,
+            it: T,
+        ) -> Result<Vec<InboundGroupSession>> {
+            Ok(it.take(limit).collect())
+        }
+
+        if let Some(requested_order) = self.get_backup_order(backup_version) {
+            // We were asked for a known backup version - only return sessions that are not
+            // already backed up in that session or a later one
+            let it =
+                self.get_inbound_group_sessions_and_backup_order().await?.into_iter().filter_map(
+                    |(session, order)| if order >= requested_order { None } else { Some(session) },
+                );
+            take_and_collect(limit, it)
+        } else {
+            // We are being asked about a new backup version - all sessions need backing up
+            let it = self.get_inbound_group_sessions().await?.into_iter();
+            take_and_collect(limit, it)
+        }
     }
 
     async fn mark_inbound_group_sessions_as_backed_up(
         &self,
-        _backup_version: &str,
+        backup_version: &str,
         room_and_session_ids: &[(&RoomId, &str)],
     ) -> Result<()> {
         for (room_id, session_id) in room_and_session_ids {
             let session = self.inbound_group_sessions.get(room_id, session_id);
             if let Some(session) = session {
                 session.mark_as_backed_up();
+
+                // Update this session's `backed_up_to` value, if the requested backup is later
+                // than the current value (or missing).
+                let requested_order = self.get_or_create_backup_order(backup_version);
+                self.inbound_group_sessions_backed_up_to
+                    .write()
```

Comment:
Probably doesn't make much of a difference but I think it would be nicer to acquire the lock once instead of acquiring it every iteration of the loop.